tests/zfs-tests/tests/functional/cp_files/cp_files_002_pos.ksh

```diff
@@ -0,0 +1,98 @@
+#! /bin/ksh -p
+#
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or https://opensource.org/licenses/CDDL-1.0.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2024 by Lawrence Livermore National Security, LLC.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+# Verify all cp --reflink modes work with modified file.
+#
+# STRATEGY:
+# 1. Verify "cp --reflink=never|auto|always" all correctly copy a
+#    freshly modified file.  Depending on the setting it may be
+#    cloned or copied but it must always be fully intact and never
+#    truncated.
+#
+
+verify_runnable "global"
+
+if ! is_linux; then
+	log_unsupported "cp --reflink is a GNU coreutils option"
+fi
+
+function cleanup
+{
+	rm -rf $TESTDIR/cp-reflink
+}
+
+function verify_copy
+{
+	src_cksum=$(sha256digest $1)
+	dst_cksum=$(sha256digest $2)
+
+	if [[ "$src_cksum" != "$dst_cksum" ]]; then
+		log_must ls -l $TESTDIR/cp
+		log_fail "checksum mismatch ($src_cksum != $dst_cksum)"
+	fi
+}
+
+log_assert "Verify all cp --reflink modes work with modified file"
+
+log_onexit cleanup
+
+RECORDSIZE=$(zfs get -Hp -o value recordsize $TESTPOOL/$TESTFS)
+SRC_FILE=src.data
+DST_FILE=dst.data
+SRC_SIZE=$(($RANDOM % 2048))
+
+log_must mkdir $TESTDIR/cp-reflink
+log_must cd $TESTDIR/cp-reflink
+
+for mode in "never" "auto" "always"; do
```

Comment:
The case where the file is partially synced is important in the reproducer because the upper layers do return an error when nothing can be cloned, and then `cp` falls back on a regular copy.

That condition is being checked here by `--reflink=always` avoiding the fallback, which looks reasonable.

That said maybe update the strategy part to describe that the always form verifies that blocks will always be cloned even when dirty?

eg. "--always forces block cloning covering the case where the file has dirty blocks"

Reply:
To make it easier to test I've added a `zfs_bclone_wait_dirty` tunable to control if the code return EINVAL right away or waits for dirty blocks.  I also went ahead and updated the test to check both modes of operation.  It'd be interesting to see for the workloads you care about what the performance difference is.

```
# STRATEGY:
# 1. Verify "cp --reflink=never|auto|always" behaves as expected.
#    Two different modes of operation are tested.
#
#    a. zfs_bclone_wait_dirty=0: FICLONE and FICLONERANGE fail with EINVAL
#       when there are dirty blocks which cannot be immediately cloned.
#       This is the default behavior.
#
#    b. zfs_bclone_wait_dirty=1: FICLONE and FICLONERANGE wait for
#       dirty blocks to be written to disk allowing the clone to succeed.
#       The downside to this is it may be slow which depending on the
#       situtation may defeat the point of making a clone.
```

Reply:
That's definitely a concern.  To make it even less likely I added a pool sync immediately before the `dd` which dirties the blocks.  While it doesn't make this impossible it should help.  We'll have to see if we get any false positives in practice here.

```ksh
       sync_pool $TESTPOOL
       log_must dd if=/dev/urandom of=$SRC_FILE bs=$RECORDSIZE seek=$SRC_SIZE \
           count=1 conv=notrunc
```